File: regression/esbmc-cpp/projects/mailutils-2.2/libmu_cpp/url.cc
/*
   GNU Mailutils -- a suite of utilities for electronic mail
   Copyright (C) 2004, 2006, 2007, 2008, 2009, 2010 Free Software
   Foundation, Inc.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 3 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General
   Public License along with this library; if not, write to the
   Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
   Boston, MA 02110-1301 USA
*/

#include <mailutils/cpp/url.h>

using namespace mailutils;

//
// Url
//

Url :: Url (const std::string& str)
{
  int status = mu_url_create (&url, str.c_str ());
  if (status)
    throw Exception ("Url::Url", status);
}

Url :: Url (const char* str)
{
  int status = mu_url_create (&url, str);
  if (status)
    throw Exception ("Url::Url", status);
}

Url :: Url (const mu_url_t url)
{
  if (url == 0)
    throw Exception ("Url::Url", EINVAL);

  this->url = url;
}

Url :: ~Url ()
{
  mu_url_destroy (&url);
}

void
Url :: parse ()
{
  int status = mu_url_parse (url);
  if (status)
    throw Exception ("Url::parse", status);
}

long
Url :: get_port ()
{
  long port;
  int status = mu_url_get_port (url, &port);
  if (status)
    throw Exception ("Url::get_port", status);
  return port;
}

std::string
Url :: get_scheme ()
{
  const char* buf = NULL;
  int status = mu_url_sget_scheme (url, &buf);
  if (status == MU_ERR_NOENT)
    return "";
  else if (status)
    throw Exception ("Url::get_scheme", status);
  return std::string (buf ? buf : "");
}

std::string
Url :: get_user ()
{
  const char* buf = NULL;
  int status = mu_url_sget_user (url, &buf);
  if (status == MU_ERR_NOENT)
    return "";
  else if (status)
    throw Exception ("Url::get_user", status);
  return std::string (buf ? buf : "");
}

Secret&
Url :: get_secret ()
{
  mu_secret_t c_secret;
  int status = mu_url_get_secret (url, &c_secret);
  if (status == MU_ERR_NOENT)
    return *new Secret ("");
  else if (status)
    throw Exception ("Url::get_secret", status);
  return *new Secret (c_secret);
}

std::string
Url :: get_auth ()
{
  const char* buf = NULL;
  int status = mu_url_sget_auth (url, &buf);
  if (status == MU_ERR_NOENT)
    return "";
  else if (status)
    throw Exception ("Url::get_auth", status);
  return std::string (buf ? buf : "");
}
 
std::string
Url :: get_host ()
{
  const char* buf = NULL;
  int status = mu_url_sget_host (url, &buf);
  if (status == MU_ERR_NOENT)
    return "";
  else if (status)
    throw Exception ("Url::get_host", status);
  return std::string (buf ? buf : "");
}

std::string
Url :: get_path ()
{
  const char* buf = NULL;
  int status = mu_url_sget_path (url, &buf);
  if (status == MU_ERR_NOENT)
    return "";
  else if (status)
    throw Exception ("Url::get_path", status);
  return std::string (buf ? buf : "");
}

std::vector<std::string>
Url :: get_query ()
{
  size_t argc;
  char **argv;
  
  int status = mu_url_sget_query (url, &argc, &argv);
  if (status)
    throw Exception ("Url::get_query", status);

  std::vector<std::string> params;

  for (int i = 0; i < argc; i++)
    params.push_back (argv[i]);

  return params;
}

std::string
Url :: to_string ()
{
  const char *str = mu_url_to_string (url);
  return std::string (str ? str : "");
}

namespace mailutils
{
  std::ostream& operator << (std::ostream& os, Url& url) {
    return os << url.to_string ();
  };
}

File: regression/esbmc-cpp/projects/mailutils-2.2/python/mailutils/url.py
#
#  GNU Mailutils -- a suite of utilities for electronic mail
#  Copyright (C) 2009, 2010 Free Software Foundation, Inc.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General
#  Public License along with this library; if not, write to the
#  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
#  Boston, MA 02110-1301 USA
#

from mailutils.c_api import url
from mailutils import secret
from mailutils.error import *

class Url:
    __owner = False

    def __init__ (self, u):
        if isinstance (u, url.UrlType):
            self.url = u
        else:
            self.url = url.UrlType ()
            self.__owner = True
            status = url.create (self.url, u)
            if status:
                raise UrlError (status)

    def __del__ (self):
        if self.__owner:
            url.destroy (self.url)
        del self.url

    def __str__ (self):
        return url.to_string (self.url)

    def parse (self):
        """Parses the url, after calling this the get functions
        can be called."""
        if self.__owner:
            status = url.parse (self.url)
            if status:
                raise UrlError (status)

    def get_port (self):
        status, port = url.get_port (self.url)
        if status:
            raise UrlError (status)
        return port

    def get_scheme (self):
        status, scheme = url.get_scheme (self.url)
        if status == MU_ERR_NOENT:
           return ''
        elif status:
            raise UrlError (status)
        return scheme

    def get_user (self):
        status, user = url.get_user (self.url)
        if status == MU_ERR_NOENT:
           return ''
        elif status:
            raise UrlError (status)
        return user

    def get_secret (self):
        status, sec = url.get_secret (self.url)
        if status == MU_ERR_NOENT:
            return secret.Secret ('')
        elif status:
            raise UrlError (status)
        return secret.Secret (sec)

    def get_auth (self):
        status, auth = url.get_auth (self.url)
        if status == MU_ERR_NOENT:
           return ''
        elif status:
            raise UrlError (status)
        return auth

    def get_host (self):
        status, host = url.get_host (self.url)
        if status == MU_ERR_NOENT:
           return ''
        elif status:
            raise UrlError (status)
        return host

    def get_path (self):
        status, path = url.get_path (self.url)
        if status == MU_ERR_NOENT:
           return ''
        elif status:
            raise UrlError (status)
        return path

    def get_query (self):
        status, query = url.get_query (self.url)
        if status == MU_ERR_NOENT:
           return ''
        elif status:
            raise UrlError (status)
        return query
